File: 216__Combination_Sum_III.py
class Solution(object):
    def combinationSum3(self, k, n):
        """
        :type k: int
        :type n: int
        :rtype: List[List[int]]
        """
        def dfs(k, n, path, result, start):
            if k == 0 and n == 0:
                result.append(path[:])
                return
            elif k < 0 or n < 0:
                return
            else:
                for i in range(start, 10):
                    if n - i < 0:
                        break
                    path.append(i)
                    dfs(k - 1, n - i, path, result, i + 1)
                    path.pop()

        result = []
        dfs(k, n, [], result, 1)
        return result

File: Java/216__Combination_Sum_III.java
class Solution {

    List<List<Integer>> result;

    public List<List<Integer>> combinationSum3(int k, int n) {
        result = new ArrayList<>();
        helper(k, n, 1, new ArrayList<>());
        return result;
    }

    private void helper(int k, int rest_n, int start, List<Integer> cur) {
        if (k == 0) {
            if (rest_n == 0) result.add(new ArrayList<>(cur));
            else return;
        }
        if (rest_n <= 0) return;
        if (start >= 10 || start > rest_n) return;
        helper(k, rest_n, start + 1, cur);
        cur.add(start);
        helper(k - 1, rest_n - start, start + 1, cur);
        cur.remove(cur.size() - 1);
    }
}
